src/c++/perf_analyzer/genai-perf/genai_perf/export_data/report_output.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+from argparse import Namespace
+
+from genai_perf.exceptions import GenAIPerfException
+from genai_perf.export_data.data_exporter_factory import (
+    DataExporterFactory,
+    DataExporterType,
+)
+from genai_perf.export_data.data_exporter_interface import DataExporterInterface
+from genai_perf.llm_metrics import LLMProfileDataParser
+from genai_perf.parser import get_extra_inputs_as_dict
+
+
+class ReportOutput:
+    """
+    A class to orchestrate output generation.
+    """
+
+    def __init__(self, data_parser: LLMProfileDataParser, args: Namespace):
+        self.data_parser = data_parser
+        self.args = args
+
+        if args.concurrency:
+            self.infer_mode = "concurrency"
+            self.load_level = f"{args.concurrency}"
+        elif args.request_rate:
+            self.infer_mode = "request_rate"
+            self.load_level = f"{args.request_rate}"
+        else:
+            raise GenAIPerfException("No valid infer mode specified")
+
+        self.stats = self.data_parser.get_statistics(self.infer_mode, self.load_level)
+
+    def report_output(self) -> None:
+        factory = DataExporterFactory()
+        data_exporters = []
+        data_exporters.append(self._get_console_exporter(factory))
+        data_exporters.append(self._get_json_exporter(factory))
+        data_exporters.append(self._get_csv_exporter(factory))
```

Comment:
I still don't see the benefit of factory here... maybe factory is the wrong name to use? Also why can't report_output.py have those information? It's pretty small file and afterall, the job of OutputReporter is to orchestrate the output generation. So why delegate that job to factory? I feel like we are adding unnecessary complexity to our codebase. 

For instance, the code below should achieve the same thing without needing the factory class nor other methods like `_get_console_exporter`:

```python
class OutputReporter:
    """
    A class to orchestrate output generation.
    """

    def __init__(self, data_parser: LLMProfileDataParser, args: Namespace):
        self.args = args

        if args.concurrency:
            self.infer_mode = "concurrency"
            self.load_level = f"{args.concurrency}"
        elif args.request_rate:
            self.infer_mode = "request_rate"
            self.load_level = f"{args.request_rate}"
        else:
            raise GenAIPerfException("No valid infer mode specified")

        self.stats = data_parser.get_statistics(self.infer_mode, self.load_level)

    def report_output(self) -> None:
        csv = CsvExporter(...)
        json = JsonExporter(...)
        console = ConsoleExporter(...)

        csv.export()
        json.export()
        console.export()
```

@tgerdesnv What are your thoughts?

Reply:
It was all in favor of future extensibility.
Your approach is functionally correct. I was trying to do the refactor, leaving a place to logically extend for potential plugins. I have not designed the plugins yet but I think my current implementation gives me a path to that solution more easily than the suggested implementation..
End of the day, if I am trying to make it ready for tomorrow but made the code base today hard to understand, I can change it.
I will wait for @tgerdesnv to chime in.

Reply:
Would like to hear from @tgerdesnv but my 2 cents is that while keeping modularity and future works in mind is definitely something we should strive for, we should also think about the readability and complexity of our codebase. This is because the purpose of modularity is to allow more easier extension and development and so does the readability and complexity of the codebase. If the new modular code makes us do _more_ work than non-modular code, I think it defeats the original purpose of making our dev life easier.

For instance, let's assume that some time in the future, we needed to add a new exporter class called `XYZExporter`. In my example above, the changes required is to add the following lines in `report_output` method:
```python
xyz = XYZExporter(
    type=DataExporterType.XYZ,
    stats=self.stats.stats_dict,
    artifact_dir=self.args.artifact_dir,
)
xyz.export()
```

On the other hand, with the current changes, the changes required are
1. Add new exporter type: `DataExporterType.XYZ`
2. Add new `DataExporterMapping` (`DataExporterType.XYZ: XYZExporter`)
1. Modify `report_output` method
```python
data_exporters.append(self._get_xyz_exporter(factory))
```

4. Add new initializer method
```python
    def _get_xyz_exporter(self, factory: DataExporterFactory) -> DataExporterInterface:
        config = {
            "type": DataExporterType.XYZ,
            "stats": self.stats.stats_dict,
            "artifact_dir": self.args.artifact_dir,
        }
        return factory.create_data_exporter(config)
```

Former only requires editing a single method in a single file, whereas the latter requires modifying multiple methods/places in two files.

Reply:
I think part of the weirdness is that output_reporter.py still needs to have all of the knowledge in the current implementation. You're not abstracting away anything. If you are going to create a factory, then it should abstract away the types and how to create them.

In my opinion, instead of 
`data_exporters.append(self._get_console_exporter(factory))`
It should be something closer to this:
`data_exporters.append(factory.get_console_exporter())`
although that still requires that the file knows the full list of exporter types. 

Full abstraction would be this:
`data_exporters = factory.get_data_exporters()`

At that point, the output_reporter.py has no knowledge of what types of exporters, how many there are, or how they are created. All of that is abstracted away. When new types are added by us or via plugins, this file wouldn't need to change.

In the absence of that abstraction, then yes there doesn't seem to be a point in the factory.

Reply:
It looks like you made this proposed change as well? I'm ok to resolve this unless Hyunjae has more comments.

Reply:
Good with the changes. Resolving the comment.